### PR TITLE
Hman 164 - Amend database mapping for relatedParty element in the POST/PUT/GET hearing end-point

### PR DESCRIPTION
**Pull request by sullivangareth**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/HMAN-164

### Change description ###

The GET, POST and PUT /hearing end-points in CftHearingService have been established as part of the Hearings Management Phase 2.

Currently, whenever the <relatedParties> element are sent in the request payload for the POST/PUT hearing, they are mapped incorrectly to the HMC database tables. This ticket is raised to change the mapping.

Similarly for the GET /hearing end-point in CftHearingService, the <relatedParties> element need to pick up the information from the correct tables.

DESCRIPTION:

1. Create new database table in the HMC database (along with the java entity objects) for the table name - PARTY_RELATIONSHIP_DETAILS. 

When a POST or PUT /hearing request is received with the <relatedParties> element in the request payload, the elements must be mapped to the database table

2.  Remove the related_party and related_party_relationship_type from the INDIVIDUAL_DETAILS table.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/hmc/data/HearingPartyEntity.java [hunk 1]
```diff
@@ -24,6 +25,7 @@
 @Table(name = "hearing_party")
 @Entity
 @Data
+@EqualsAndHashCode(exclude = "organisationDetailEntity")
```
Comment (sullivangareth) [inline src/main/java/uk/gov/hmcts/reform/hmc/data/HearingPartyEntity.java:28]: Had to add this to break hashCode cyclic dependency that caused `StackOverFlowError`

File: src/main/java/uk/gov/hmcts/reform/hmc/data/HearingPartyEntity.java [hunk 2]
```diff
@@ -68,4 +70,6 @@ public class HearingPartyEntity {
     @OneToMany(mappedBy = "hearingParty", cascade = CascadeType.PERSIST, orphanRemoval = true)
     private List<ContactDetailsEntity> contactDetails;
 
+    @OneToMany(cascade = CascadeType.PERSIST, orphanRemoval = true)
```
Comment (sullivangareth) [inline src/main/java/uk/gov/hmcts/reform/hmc/data/HearingPartyEntity.java:73]: Not convinced this is correct, but the `mappedBy = hearingParty` caused issues that stopped AppContext from starting

File: src/main/java/uk/gov/hmcts/reform/hmc/data/PartyRelationshipDetailsEntity.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import javax.persistence.Id;
+import javax.persistence.Table;
+
+@Table(name = "party_relationship_details")
+@Entity
+@Data
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class PartyRelationshipDetailsEntity implements Serializable {
```
Comment (sullivangareth) [inline src/main/java/uk/gov/hmcts/reform/hmc/data/PartyRelationshipDetailsEntity.java:22]: Does this need some foreign key/table link ?

File: src/main/java/uk/gov/hmcts/reform/hmc/data/PartyRelationshipDetailsEntity.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY,
+            generator = "party_relationship_details_id_seq")
+    @Column(name = "party_relationship_details_id")
+    private Long partyRelationshipDetailsId;
+
+    @Column(name = "source_tech_party_id")
+    private Long sourceTechPartyId;
```
Comment (sullivangareth) [inline src/main/java/uk/gov/hmcts/reform/hmc/data/PartyRelationshipDetailsEntity.java:31]: I think these could be the relevant entity objects, but couldn't get it working

File: src/main/java/uk/gov/hmcts/reform/hmc/helper/GetHearingResponseMapper.java [hunk 5]
```diff
@@ -30,14 +31,23 @@
...
 
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Optional;
+import java.util.stream.Collectors;
 
 @Component
 public class GetHearingResponseMapper extends GetHearingResponseCommonCode {
 
+    private HearingPartyRepository hearingPartyRepository;
```
Comment (sullivangareth) [inline src/main/java/uk/gov/hmcts/reform/hmc/helper/GetHearingResponseMapper.java:44]: This feels all kinds of wrong to have a repository within a mapper ?

File: src/main/java/uk/gov/hmcts/reform/hmc/repository/HearingPartyRepository.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+import uk.gov.hmcts.reform.hmc.data.HearingPartyEntity;
+
+@Transactional(propagation = Propagation.REQUIRED)
+@Repository
+public interface HearingPartyRepository extends JpaRepository<HearingPartyEntity, Long> {
+
+    @Query("SELECT techPartyId from HearingPartyEntity where partyReference = :partyRef")
+    Long getTechPartyIdByReference(String partyRef);
+
+    @Query("SELECT partyReference from HearingPartyEntity where techPartyId = :techPartyId")
```
Comment (sullivangareth) [inline src/main/java/uk/gov/hmcts/reform/hmc/repository/HearingPartyRepository.java:17]: This "reverse lookup" is needed to workout, given the values in `party_relationship_details` table, the values tor return in the "GET" response

File: src/main/java/uk/gov/hmcts/reform/hmc/repository/PartyRelationshipDetailsRepository.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
+import uk.gov.hmcts.reform.hmc.data.PartyRelationshipDetailsEntity;
+
+import java.util.List;
+
+@Transactional(propagation = Propagation.REQUIRED)
+@Repository
+public interface PartyRelationshipDetailsRepository extends CrudRepository<PartyRelationshipDetailsEntity, Long> {
```
Comment (sullivangareth) [inline src/main/java/uk/gov/hmcts/reform/hmc/repository/PartyRelationshipDetailsRepository.java:13]: Not sure if this is needed now - think it was for testing.  Will check and remove if not needed

Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/hmc/repository/PartyRelationshipDetailsRepository.java:11]: Removed in https://github.com/hmcts/hmc-cft-hearing-service/pull/153/commits/e430840b5052c4fef7fcddd4e3f82f9670491048

File: src/main/java/uk/gov/hmcts/reform/hmc/data/HearingPartyEntity.java (hunk 2 above)
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/hmc/data/HearingPartyEntity.java]: Fixed by @danlysiak in https://github.com/hmcts/hmc-cft-hearing-service/pull/153/commits/50b833069211b0402b67717fae973ed563286907

File: src/main/java/uk/gov/hmcts/reform/hmc/data/PartyRelationshipDetailsEntity.java (hunk 3 above)
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/hmc/data/PartyRelationshipDetailsEntity.java]: Fixed by @danlysiak in https://github.com/hmcts/hmc-cft-hearing-service/pull/153/commits/50b833069211b0402b67717fae973ed563286907

File: src/main/java/uk/gov/hmcts/reform/hmc/data/PartyRelationshipDetailsEntity.java (hunk 4 above)
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/hmc/data/PartyRelationshipDetailsEntity.java]: Fixed by @danlysiak in https://github.com/hmcts/hmc-cft-hearing-service/pull/153/commits/50b833069211b0402b67717fae973ed563286907

File: src/main/java/uk/gov/hmcts/reform/hmc/helper/GetHearingResponseMapper.java (hunk 5 above)
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/hmc/helper/GetHearingResponseMapper.java]: Removed, Fixed by @danlysiak in https://github.com/hmcts/hmc-cft-hearing-service/pull/153/commits/50b833069211b0402b67717fae973ed563286907

File: src/main/java/uk/gov/hmcts/reform/hmc/repository/HearingPartyRepository.java (hunk 6 above)
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/hmc/repository/HearingPartyRepository.java]: Both native queries now removed, Fixed by @danlysiak in https://github.com/hmcts/hmc-cft-hearing-service/pull/153/commits/50b833069211b0402b67717fae973ed563286907

File: src/main/java/uk/gov/hmcts/reform/hmc/service/HearingManagementServiceImpl.java [hunk 8]
```diff
@@ -259,6 +270,23 @@ private HearingEntity saveHearingDetails(HearingRequest createHearingRequest) {
         return hearingRepository.save(hearingEntity);
     }
 
+    private void savePartyRelationshipDetails(HearingRequest createHearingRequest, HearingEntity hearingEntity) {
+
+        final List<PartyDetails> partyDetails = createHearingRequest.getPartyDetails();
+        final List<HearingPartyEntity> hearingParties = hearingEntity.getLatestCaseHearingRequest().getHearingParties();
```
Comment (sullivangareth) [inline src/main/java/uk/gov/hmcts/reform/hmc/service/HearingManagementServiceImpl.java:276]: Could `hearingParties` be null here ? If so what would we do - just not save any partyRelationshipDetails to DB ?

File: src/main/java/uk/gov/hmcts/reform/hmc/data/HearingPartyEntity.java (hunk 1 above)
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/hmc/data/HearingPartyEntity.java]: Removed in [204b934](https://github.com/hmcts/hmc-cft-hearing-service/pull/153/commits/204b934d37f2d1ffc62ac4080bd9e718ed5234dc)

File: src/main/java/uk/gov/hmcts/reform/hmc/service/HearingManagementServiceImpl.java (hunk 8 above)
Comment (danlysiak) [reply src/main/java/uk/gov/hmcts/reform/hmc/service/HearingManagementServiceImpl.java:300]: There can be no parties according to the spec, and relationships are submitted against parties in the request. We should do nothing in this scenario. On line 302 we could harden it by:
```
if (!Collections.isEmpty(partyDetails) && !Collections.isEmpty(hearingParties))
```
(Note the `isEmpty()` util method also checks for null))

File: src/main/java/uk/gov/hmcts/reform/hmc/service/HearingManagementServiceImpl.java [hunk 9]
```diff
@@ -283,6 +294,23 @@ private HearingEntity saveHearingDetails(HearingRequest createHearingRequest) {
...
+
+        if (partyDetails != null) {
+            for (PartyDetails partyDetail : partyDetails) {
+                if (partyDetail.getIndividualDetails() != null
+                    && partyDetail.getIndividualDetails().getRelatedParties() != null) {
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/hmc/service/HearingManagementServiceImpl.java:305]:
```suggestion
                    && !Collections.isEmpty(partyDetail.getIndividualDetails().getRelatedParties()) {
```

File: src/main/resources/db/migration/V20220413_164.2__HMAN-164-CreatePartyRelationshipDetails.sql [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    START WITH 1
+    INCREMENT BY 1
+    NO MINVALUE
+    NO MAXVALUE
+    CACHE 1;
+
+CREATE TABLE public.party_relationship_details (party_relationship_details_id bigint not null default nextval('public.party_relationship_details_id_seq'::regclass),
+                                             source_tech_party_id bigint not null,
+                                             target_tech_party_id bigint not null,
+                                             relationship_type varchar(2000)
```
Comment (danlysiak) [inline src/main/resources/db/migration/V20220413_164.2__HMAN-164-CreatePartyRelationshipDetails.sql:11]:
```suggestion
                                             relationship_type varchar(10)
```
This was supposed to be the new length in HMAN-90

File: src/main/java/uk/gov/hmcts/reform/hmc/helper/PartyRelationshipDetailsMapper.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+        if (relatedParties != null) {
+            for (RelatedParty relatedParty: relatedParties) {
+
+                String relatedPartyId = relatedParty.getRelatedPartyID();
+
+                final HearingPartyEntity targetTechParty =
+                        getHearingPartyEntityByReference(relatedPartyId, hearingPartyEntities);
+
+                final HearingPartyEntity sourceTechParty =
+                        getHearingPartyEntityByReference(partyDetails.getPartyID(), hearingPartyEntities);
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/hmc/helper/PartyRelationshipDetailsMapper.java:31]: Could move this line above this `for` loop (still inside the `if`) so we only call it once.

File: src/main/java/uk/gov/hmcts/reform/hmc/service/HearingManagementServiceImpl.java (hunk 8 above)
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/hmc/service/HearingManagementServiceImpl.java:289]: Fixed in https://github.com/hmcts/hmc-cft-hearing-service/pull/153/commits/768e22b379256ea506790b5fc818777349748c05

File: src/main/java/uk/gov/hmcts/reform/hmc/service/HearingManagementServiceImpl.java (hunk 9 above)
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/hmc/service/HearingManagementServiceImpl.java]: Fixed in https://github.com/hmcts/hmc-cft-hearing-service/pull/153/commits/768e22b379256ea506790b5fc818777349748c05

File: src/main/java/uk/gov/hmcts/reform/hmc/helper/PartyRelationshipDetailsMapper.java (hunk 11 above)
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/hmc/helper/PartyRelationshipDetailsMapper.java]: Fixed in https://github.com/hmcts/hmc-cft-hearing-service/pull/153/commits/768e22b379256ea506790b5fc818777349748c05

File: src/main/java/uk/gov/hmcts/reform/hmc/helper/PartyRelationshipDetailsMapper.java [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        }
+
+        return partyRelationshipDetails;
+    }
+
+    private HearingPartyEntity getHearingPartyEntityByReference(String relatedPartyId,
+                                                                List<HearingPartyEntity> hearingPartyEntities) {
+        return hearingPartyEntities.stream()
+                .filter(hearingPartyEntity -> hearingPartyEntity.getPartyReference().equals(relatedPartyId))
+                .findFirst()
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/hmc/helper/PartyRelationshipDetailsMapper.java]: Fixed in https://github.com/hmcts/hmc-cft-hearing-service/pull/153/commits/3386b7b01d777d649dc0072208657cee873969dd

File: src/main/resources/db/migration/V20220413_164.2__HMAN-164-CreatePartyRelationshipDetails.sql (hunk 10 above)
Comment (sullivangareth) [reply src/main/resources/db/migration/V20220413_164.2__HMAN-164-CreatePartyRelationshipDetails.sql]: Fixed in https://github.com/hmcts/hmc-cft-hearing-service/pull/153/commits/165c047e535d05e348e3cb68bb5289bac865701a

**Review by danlysiak — APPROVED** [review]

**Pull request merged**